### DISCUSSION TITLE
TypeError: source must an AudioSource not NoneType

### DISCUSSION THREAD

**Discussion by mrracoonman** [Rapptz/discord.py#6071] [Q&A] [answered]
_I'm attempting to get a bot to play audio in a voice channel, this is the code I have for it._ 

@bot.command(pass_context=True)
async def play(ctx, url: str):
    song_there = os.path.isfile('song.mp3')
    try:
        if song_there:
            os.remove('song.mp3')
            print('Removed old song file.')
    except PermissionError:
        print('Trying to delete song file, but is being played.')
        await ctx.send(f'ERROR, Music playing')
        return

    await ctx.send(f'Getting everything ready now')

    voice = ctx.guild.voice_client
    ydl_opts = {
        'format': 'bestaudio/best',
        'postprocessors': [{
            'key': 'FFmpegExtractAudio',
            "preferredcodec": 'mp3',

        }],
    }

    with youtube_dl.YoutubeDL(ydl_opts) as ydl:
        print(f'Downloading audio now')
        ydl.download([url])

    for file in os.listdir('./'):
        if file.endswith('.mp3'):
            name = file
            print(f'Renamed file: {file}')
            os.rename(file, 'song.mp3')

    voice.play(discord.ffmpegPCMaudio('song.mp3'), after=lambda e: print(f'{name} has finished playing'))
    voice.source = discord.PCMVolumeTransformer(voice.source)
    voice.source.volume = 0.2

    nname = name.rsplit('-', 2)
    await ctx.send(f'Playing {nname}')
    print(f'Playing')

_And it works until about somewhere in here:_

  for file in os.listdir('./'):
        if file.endswith('.mp3'):
            name = file
            print(f'Renamed file: {file}')
            os.rename(file, 'song.mp3')

    voice.play(discord.ffmpegPCMaudio('song.mp3'), after=lambda e: print(f'{name} has finished playing'))
    voice.source = discord.PCMVolumeTransformer(voice.source)
    voice.source.volume = 0.2

_Then it sends this error._

Ignoring exception in command play:
Traceback (most recent call last):
  File "C:\Users\User\PycharmProjects\pythonProject\venv\lib\site-packages\discord\ext\commands\core.py", line 85, in wrapped
    ret = await coro(*args, **kwargs)
  File "C:/Users/User/PycharmProjects/pythonProject/main.py", line 136, in play
    voice.play(discord.ffmpegPCMaudio('song.mp3'), after=lambda e: print(f'{name} has finished playing'))
  File "C:\Users\User\PycharmProjects\pythonProject\venv\lib\site-packages\discord\voice_client.py", line 558, in play
    raise TypeError('source must an AudioSource not {0.__class__.__name__}'.format(source))
TypeError: source must an AudioSource not NoneType

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\User\PycharmProjects\pythonProject\venv\lib\site-packages\discord\ext\commands\bot.py", line 903, in invoke
    await ctx.command.invoke(ctx)
  File "C:\Users\User\PycharmProjects\pythonProject\venv\lib\site-packages\discord\ext\commands\core.py", line 859, in invoke
    await injected(*ctx.args, **ctx.kwargs)
  File "C:\Users\User\PycharmProjects\pythonProject\venv\lib\site-packages\discord\ext\commands\core.py", line 94, in wrapped
    raise CommandInvokeError(exc) from exc
discord.ext.commands.errors.CommandInvokeError: Command raised an exception: TypeError: source must an AudioSource not NoneType

_I cant figure out why it's displaying said error_

_Win 10, py 3.8.6_

**Comment by Harmon758** [comment, marked as answer]
Your traceback would seem to indicate that you're passing `None` to `voice.play`, but attempting to access `discord.ffmpegPCMaudio` itself should error, as the class is [`FFmpegPCMAudio`](https://discordpy.readthedocs.io/en/latest/api.html#discord.FFmpegPCMAudio).

For code block usage, see https://docs.github.com/en/free-pro-team@latest/github/writing-on-github/creating-and-highlighting-code-blocks.
